src/client/builtin_scene/gaussian_splats_mesh.cpp

```diff
@@ -0,0 +1,189 @@
+      : meshes::Splat()
+      , needsRebuild_(false)
+      , needsSorting_(false)
+      , bufferInitialized_(false)
+  {
+  }
+
+  void GaussianSplatsMesh::addSplatsEntity(ecs::EntityId entityId)
+  {
+    splatEntities_.insert(entityId);
+    needsRebuild_ = true;
+    needsSorting_ = true;
+  }
+
+  void GaussianSplatsMesh::removeSplatsEntity(ecs::EntityId entityId)
+  {
+    auto it = splatEntities_.find(entityId);
+    if (it != splatEntities_.end())
+    {
+      splatEntities_.erase(it);
+      needsRebuild_ = true;
+      needsSorting_ = true;
+    }
+  }
+
+  void GaussianSplatsMesh::updateSplatsEntity(ecs::EntityId entityId)
+  {
+    if (splatEntities_.count(entityId) > 0)
+    {
+      needsRebuild_ = true;
+      needsSorting_ = true;
+    }
+  }
+
+  void GaussianSplatsMesh::setupSplatBuffer(std::shared_ptr<client_graphics::WebGL2Context> glContext,
+                                            std::shared_ptr<client_graphics::WebGLVertexArray> vao)
+  {
+    if (!glContext || bufferInitialized_)
+      return;
+
+    // Create the instance buffer for splat data
+    splatInstanceBuffer_ = glContext->createBuffer();
+
+    if (splatInstanceBuffer_)
+    {
+      bufferInitialized_ = true;
+    }
+  }
+
+  void GaussianSplatsMesh::updateSplatBuffer(std::shared_ptr<client_graphics::WebGL2Context> glContext)
+  {
+    if (!glContext || !bufferInitialized_ || sortedSplats_.empty())
+      return;
+
+    // Prepare a contiguous array of GPU data (excluding depth and sourceEntity)
+    std::vector<float> gpuData;
+    size_t floatsPerSplat = SPLAT_STRIDE / sizeof(float);
+    gpuData.reserve(sortedSplats_.size() * floatsPerSplat);
+
+    for (const auto &splat : sortedSplats_)
+    {
+      // Position (vec3)
+      gpuData.push_back(splat.position.x);
+      gpuData.push_back(splat.position.y);
+      gpuData.push_back(splat.position.z);
+
+      // Color (vec3)
+      gpuData.push_back(splat.color.r);
+      gpuData.push_back(splat.color.g);
+      gpuData.push_back(splat.color.b);
+
+      // Opacity (float)
+      gpuData.push_back(splat.opacity);
+
+      // Scale (vec3)
+      gpuData.push_back(splat.scale.x);
+      gpuData.push_back(splat.scale.y);
+      gpuData.push_back(splat.scale.z);
+
+      // Rotation (vec4)
+      gpuData.push_back(splat.rotation.x);
+      gpuData.push_back(splat.rotation.y);
+      gpuData.push_back(splat.rotation.z);
+      gpuData.push_back(splat.rotation.w);
+    }
+
+    // Upload to GPU
+    glContext->bindBuffer(client_graphics::WebGLBufferBindingTarget::kArrayBuffer, splatInstanceBuffer_);
+    glContext->bufferData(client_graphics::WebGLBufferBindingTarget::kArrayBuffer,
+                          gpuData.size() * sizeof(float),
+                          gpuData.data(),
+                          client_graphics::WebGLBufferUsage::kDynamicDraw);
+  }
+
+  void GaussianSplatsMesh::setupSplatAttributes(std::shared_ptr<client_graphics::WebGLProgram> program,
```

Comment:
Reference SceneRenderer::configureMeshVertexData(). You need to abstract GaussianSplatsMesh and InstancedMesh so that they both configure instance attributes in the same way.